diff --git a/core/Makefile b/core/Makefile
index b8c4189..0255648 100644
--- a/core/Makefile
+++ b/core/Makefile
@@ -6921,12 +6921,14 @@ ifneq ($(PRODUCT_SYSTEM_DLKM_BASE_FS_PATH),)
 	$(hide) cp $(PRODUCT_SYSTEM_DLKM_BASE_FS_PATH) \
 	  $(zip_root)/META/$(notdir $(PRODUCT_SYSTEM_DLKM_BASE_FS_PATH))
 endif
+ifneq ($(TARGET_PRESERVE_EXTERNAL_RECOVERY),true)
 ifeq ($(TARGET_OTA_ALLOW_NON_AB),true)
 ifneq ($(INSTALLED_RECOVERYIMAGE_TARGET),)
 	$(hide) PATH=$(INTERNAL_USERIMAGES_BINARY_PATHS):$$PATH MKBOOTIMG=$(MKBOOTIMG) \
 	    $(MAKE_RECOVERY_PATCH) $(zip_root) $(zip_root)
 endif
 endif
+endif
 ifeq ($(AB_OTA_UPDATER),true)
 	@# When using the A/B updater, include the updater config files in the zip.
 	$(hide) cp $(TOPDIR)system/update_engine/update_engine.conf $(zip_root)/META/update_engine_config.txt
