diff --git a/sepolicy/vendor/file_contexts b/sepolicy/vendor/file_contexts
index 2c84640..ab2cfa4 100644
--- a/sepolicy/vendor/file_contexts
+++ b/sepolicy/vendor/file_contexts
@@ -53,7 +53,6 @@
 /dev/mtp_usb*                u:object_r:mtp_device:s0
 /dev/s5p-smem                u:object_r:secmem_device:s0
 /dev/rfkill                  u:object_r:rfkill_device:s0
-/dev/sec-nfc                 u:object_r:nfc_device:s0
 /dev/seiren                  u:object_r:seiren_device:s0
 /dev/stune(/.*)?             u:object_r:cgroup:s0
 /dev/v4l-subdev[0-9]*        u:object_r:video_device:s0
